File: primer3/primer3.js
function createLinkedList(posts) {
    // Validate posts
    if (!Array.isArray(posts) || posts.length === 0) {
      return null;
    }
  
    // Validate each post
    posts.forEach(post => {
      // More specific order of validations
      if (post.text === undefined || post.timestamp === undefined || post.author === undefined) {
        throw new Error('Post must contain text, timestamp, and author');
      }
  
      // Validate text first
      if (post.text === null || typeof post.text !== 'string') {
        throw new Error('Post text must be a string');
      }
      if (post.text.trim() === '') {
        throw new Error('Post text cannot be empty');
      }
  
      // Validate timestamp
      try {
        const date = new Date(post.timestamp);
        if (isNaN(date.getTime())) {
          throw new Error('Invalid timestamp format');
        }
      } catch {
        throw new Error('Invalid timestamp format');
      }
  
      // Validate author
      if (post.author === null || typeof post.author !== 'string') {
        throw new Error('Author must be a string');
      }
      if (post.author.trim() === '') {
        throw new Error('Author name cannot be empty');
      }
    });
  
    // Create linked list
    const head = { 
      data: posts[0], 
      next: null 
    };
    let current = head;
  
    for (let i = 1; i < posts.length; i++) {
      current.next = { 
        data: posts[i], 
        next: null 
      };
      current = current.next;
    }
  
    return head;
  }
  
  function searchSocialMediaFeed(feed, keyword) {
    if (!feed) return [];
  
    const results = [];
    const normalisedKeyword = keyword.toLowerCase();
    
    let current = feed;
    while (current) {
      const normalisedText = current.data.text.toLowerCase();
      
      // Split into words to handle more granular partial matches
      const textWords = normalisedText.split(/\s+/);
      const keywordWords = normalisedKeyword.split(/\s+/);
  
      // Check if any word partially matches
      const hasPartialMatch = textWords.some(word => 
        textWords.some(textWord => textWord.includes(normalisedKeyword)) || 
        keywordWords.some(keyWord => word.includes(keyWord))
      );
  
      if (hasPartialMatch) {
        results.push(current.data);
      }
      
      current = current.next;
    }
  
    return results;
  }
  
  // For compatibility with tests
  const Node = function() {};
  
  module.exports = { createLinkedList, searchSocialMediaFeed, Node };

File: primer3/primer3.py
from datetime import datetime

def validate_post(post):
    required_fields = {"text", "timestamp", "author"}
    if not all(field in post for field in required_fields):
        raise ValueError("Post missing required fields")
    
    if not post["text"] or not isinstance(post["text"], str):
        raise ValueError("Post text must be a non-empty string")
    
    if not post["author"] or not isinstance(post["author"], str):
        raise ValueError("Author must be a non-empty string")
    
    try:
        datetime.strptime(post["timestamp"], "%Y-%m-%d %H:%M:%S")
    except ValueError:
        raise ValueError("Invalid timestamp format. Use YYYY-MM-DD HH:MM:SS")

def create_linked_list(posts):
    if not posts:
        return None
    
    # Validate all posts first
    for post in posts:
        validate_post(post)
    
    # Create the head node
    head = {"data": posts[0], "next": None}
    current = head
    
    # Add remaining posts
    for post in posts[1:]:
        current["next"] = {"data": post, "next": None}
        current = current["next"]
    
    return head

def search_social_media_feed(feed, keyword):
    
    if not feed or not keyword:
        return []
    
    results = []
    keyword = keyword.lower()
    current = feed
    
    while current:
        if keyword in current["data"]["text"].lower():
            results.append(current["data"])
        current = current["next"]
    
    return results